File: Kotlin/tests/GameTest.kt
class GameTest() {

    fun run() {
        followsRuleForBlackCell()
        followsRuleForWhiteCell()
    }

    fun followsRuleForBlackCell() {
        var board = Board(array(Position(0,0)).toSet())
        var ant = Ant(Position(0,0), Direction.North)
        var game = Game(board, ant)

        game.tick()

        assert(ant.direction==Direction.West, "ant has wrong orientation")
        assert(ant.position==Position(-1, 0), "ant has wrong position")
        assert(!board.isBlack(Position(0,0)), "color of the cell should be flipped")
    }

    fun followsRuleForWhiteCell() {
        var board = Board(array(Position(0,0)).toSet())
        var ant = Ant(Position(1,1), Direction.North)
        var game = Game(board, ant)

        game.tick()

        assert(ant.direction==Direction.East, "ant has wrong orientation")
        assert(ant.position==Position(2,1), "ant has wrong position")
        assert(board.isBlack(Position(1,1)), "color of the cell should be flipped")
    }
}

File: Dart/test/GameTest.dart
class GameTest {
  run() {
    
    group('Game.tick', () {
      
      test('follows rules on black cell', () {
        Cell cell = new Cell(0,0);
        Ant ant = new Ant(new Direction.north(), cell);
        Board board = new Board([cell]);
        Game game = new Game(ant, board);
        
        game.tick();
        
        Expect.equals(new Direction.west(), ant.direction);
        Expect.equals(new Cell(-1,0), ant.position);
        Expect.isFalse(board.isBlack(cell));
      });
      
      test('follows rules on white cell', () {
        Cell cell = new Cell(0,0);
        Ant ant = new Ant(new Direction.north(), cell);
        Board board = new Board([]);
        Game game = new Game(ant, board);
        
        game.tick();
        
        Expect.equals(new Direction.east(), ant.direction);
        Expect.equals(new Cell(1,0), ant.position);
        Expect.isTrue(board.isBlack(cell));
      });
    });
  }
}